broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/NewPartitionTransitionImplTest.java

```diff
@@ -81,75 +80,90 @@ void shouldCallTransitionStepsInOrder() {
   }
 
   @Test
-  void shouldAbortTransitionIfOneStepThrowsAnException() {
+  // regression test for https://github.com/camunda-cloud/zeebe/issues/7873
+  void shouldNotStartMultipleTransitions() {
     // given
-    final var testException = new Exception("TEST_EXCEPTION");
-    when(mockStep1.transitionTo(mockContext, DEFAULT_TERM, DEFAULT_ROLE))
-        .thenReturn(TEST_CONCURRENCY_CONTROL.failedFuture(testException));
-    when(mockStep2.transitionTo(mockContext, DEFAULT_TERM, DEFAULT_ROLE))
+    final var firstStepFirstTransitionFuture = TEST_CONCURRENCY_CONTROL.<Void>createFuture();
+    final var firstStepSecondTransitionFuture = TEST_CONCURRENCY_CONTROL.<Void>createFuture();
+    final var firstStepThirdTransitionFuture = TEST_CONCURRENCY_CONTROL.<Void>createFuture();
+    when(mockStep1.transitionTo(any(), anyLong(), any()))
+        .thenReturn(firstStepFirstTransitionFuture)
+        .thenReturn(firstStepSecondTransitionFuture)
+        .thenReturn(firstStepThirdTransitionFuture);
+    when(mockStep1.prepareTransition(any(), anyLong(), any()))
         .thenReturn(TEST_CONCURRENCY_CONTROL.completedFuture(null));
 
-    final var sut = new NewPartitionTransitionImpl(of(mockStep1, mockStep2), mockContext);
-    sut.setConcurrencyControl(TEST_CONCURRENCY_CONTROL);
+    final var transition = new NewPartitionTransitionImpl(of(mockStep1), mockContext);
+    transition.setConcurrencyControl(TEST_CONCURRENCY_CONTROL);
+
+    final var firstTransitionFuture = transition.transitionTo(1, Role.FOLLOWER);
 
     // when
-    final var actualResult = sut.transitionTo(DEFAULT_TERM, DEFAULT_ROLE);
+    transition.transitionTo(2, Role.LEADER);
+    transition.transitionTo(2, Role.FOLLOWER);
+
+    firstStepFirstTransitionFuture.complete(null);
+    await().until(firstTransitionFuture::isDone);
 
     // then
-    verify(mockStep2, never()).transitionTo(mockContext, DEFAULT_TERM, DEFAULT_ROLE);
+    final var inOrder = inOrder(mockStep1);
 
-    assertThatThrownBy(actualResult::join)
-        .isInstanceOf(CompletionException.class)
-        .getCause()
-        .isSameAs(testException);
+    // first transition sequence
+    inOrder.verify(mockStep1).onNewRaftRole(mockContext, Role.FOLLOWER);
+    inOrder.verify(mockStep1).transitionTo(mockContext, 1, Role.FOLLOWER);
+
+    // second transition
+    inOrder.verify(mockStep1).onNewRaftRole(mockContext, Role.LEADER);
+    inOrder.verify(mockStep1).prepareTransition(mockContext, 2, Role.LEADER);
+    inOrder.verify(mockStep1).transitionTo(mockContext, 2, Role.LEADER);
+
+    // when reaching wait state (future) nothing more should happen
+    inOrder.verifyNoMoreInteractions();
   }
 
   @Test
-  void shouldAbortOngoingTransitionWhenNewTransitionIsRequested() {
+  // regression test for https://github.com/camunda-cloud/zeebe/issues/7873
+  void shouldExecuteTransitionsInOrder() {
```

Comment:
Would be nice to have one more test  - shouldNotExecuteFollowingTransitionsIfOneFails